File: Javascript/Algorithms/102.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
let levelOrder = (root) => {
  if (!root) return [];

  const stack = [[root, 0]], results = [];
  while (stack.length > 0) {
    let [node, level] = stack.pop();

    if (level >= results.length)
      results.push([node.val]);
    else
      results[level].push(node.val);

    let left = node.left, right = node.right;
    if (right) stack.push([right, level + 1]);
    if (left) stack.push([left, level + 1]);
  }
  return results;
};

File: Python/Algorithms/102.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root):
        if not root:
            return []

        stack, results = [(root, 0)], []
        while stack:
            node, level = stack.pop()
            left, right, val = node.left, node.right, node.val
            if level < len(results):
                results[level].append(val)
            else:
                results.append([val])
            level += 1
            if right:
                stack.append((right, level))
            if left:
                stack.append((left, level))
        return results
